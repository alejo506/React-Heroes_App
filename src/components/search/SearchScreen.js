import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router';
import { heroes } from '../../data/heroes'
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {//1.QueryString. Hacemos el history para poder hacer un push()

    //3.QueryString. Utilizamos el hook para extraer la "location "de searchScreen, esta "location" se pudo hacer igual al history, pero, es solo para ver el ejemplo utilizando el customHook
    const location = useLocation();
    // console.log(location); //location.search para imprimir la propiedad search de la location
    //4.QueryString. Uso de npm query-string
    // console.log(location.search);
    // console.log(queryString);
    //Pasamos nuestro objeto location.search a queryString
    // console.log(queryString.parse(location.search));
    //5.QueryString. Desestructuramos para extraer solo el query que nos interesa el cual es solo uno en este caso solo utilizaremos el query "q"
    const { q = '' } = queryString.parse(location.search);//Inicializamos el query en como string vacío para que no sea undefined en caso de que no tenga un valor
    // console.log(q); //Imprime batman

    //----------------------------------------------------------------------------------------------------------------------------------------------

    //Uso del customHook useForm
    const [values, handledInputChange,reset] = useForm({ //Primero necesitamos el name del input
        // searchText:''. Lo incializamos como un string vacío
        searchText: q // 6.QueryString. Agregamos como valor inicial de nuestro searchText el query(q), de esta forma si se realiza un refresh de nuestra pagina, se mantendrá como valor incicial el "q", por lo tanto muestra el superheroe que se encuentra en el query
    });


    //----------------------------------------------------------------------------------------------------------------------------------------------

    //Segundo paso, extraemos el searchText de nuestro values. Hacemos la desestructuracion.
    //Utilizamos el searchText que extraimos desde values para usarlo en el form(en values="searchText")
    const { searchText } = values;
    //----------------------------------------------------------------------------------------------------------------------------------------------

    //Creamos una costante para almacenar los heroes y luego realizar el map()
    // const heroesFiltered = heroes;
    //Utilizamos la función para obtener los heroes filtrados. Video 182
    // Utilizamos el useMemo porque cada vez que digitamos una tecla del teclado vuelve a ejecutar esta función, entonces, ahora solo se ejecuta cuando cambie el query(q)
    // const heroesFiltered = useMemo(() => getHeroesByName( searchText ), [q]); //El searchText representa al "name" en el archivo getHeroesByName
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]); //Si lo agrego como la linea de arriba, la linea 42 va a funcionar, pero, muestra una advertencia "React useMemo has a missing dependency": 'searchText'", eso quiere decir que useMemo tiene una dependencia faltante debido a que searchText tendría que estar como dependencia en lugar de la "q" 
    //Tambien gracias al useMemo si le damos click a un personaje y luego le damos en el boton "Back" me muestra la busqueda tal y como se realizó la última vez, ya que, estamos trabajando en base al Query String y la funcíon getHeroesByName no se está disparando gracias al useMemo porque tenemos memorizados esos valores

    //----------------------------------------------------------------------------------------------------------------------------------------------

    
    const handleSearch = (e) => {
        e.preventDefault();//Para que NO haga el refresh de la pagina
        // console.log(searchText);//Imprimimos lo que sea que la persona escriba y damos a la tecla "Enter"

        //2.Query String. Cuando la persona ingresa al texto(el que imprimimos en consola(searchText)), hacemos el push()
        history.push(`?q=${searchText}`);//Añadimos una "?q" de query(se puede agregar otra cosa en lugar de la "q") y luego agregamos el texto que buscamos(serachText);
        
    }
    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className="col-10">

                    <h4>Search form</h4>
                    <hr />


                    <form onSubmit={handleSearch} className="d-flex">

                        <input
                            type="search"
                            placeholder="Find your heroe"
                            className="form-control me-2"
                            aria-label="Search"
                            autoComplete="off"
                            name="searchText"
                            values={searchText}
                            onChange={handledInputChange}
                        />
                        <button
                            type="submit" /*Esto es para cuando le demos click al boton o teclemos Enter entonces que haga el posteo */
                            className="btn m-1 btn-block btn-outline-primary"
                        >
                            Search...
                    </button>
                    </form>


                </div>



            </div>
            {/* Aquí colocamos el resultado de un arreglo */}

            <h4 className="mt-4"> Results </h4>
            <hr />

                {
                    (q === '') 
                    &&
                    <div className="text-center alert alert-info">
                        Search a hero
                    </div>
                }

                {
                    (q !== '' && heroesFiltered.length === 0) //Si el query es distinto de vacío y si la longitud del heroesFiltered es 0 entonces haga lo siguiente: 
                    &&
                    <div className="text-center alert alert-danger">
                        There is no a hero with <strong>{q}</strong> 
                    </div>
                }
            <div className="row row-cols-auto row-cols-lg-3 g-auto g-lg-3 row-cols-xl-4 g-xl-auto row-cols-xs-2 g-xs-2 animate__animated animate__fadeIn">
                
                
                {/* Colocamos la expression de JS */}
                {
                    heroesFiltered.map(herOe => ( //El parentesis ver es porque retorna un objeto

                        <HeroCard
                            key={herOe.id}
                            //Eviamos todas las propiedades del heroe
                            {...herOe}
                        />
                    ))
                }
            </div>
        </div>
    )
}
