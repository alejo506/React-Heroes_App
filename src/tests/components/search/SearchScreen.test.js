import React from 'react';
import {mount} from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas en <SearchScreen />', () => {

    test('debe de mostrarse correctamente con valores por defecto ', () => {
        
        const wrapper= mount(

            //Como vamos a trabajar con rutas utilizamos el MemoryRouter
            <MemoryRouter initialEntries={['/search']}> 
                {/*Trabajamos con rutas porque cuando escribimos algo en el input y le damos enter trabajamos con el url y estar leyendolo, por esa razón vamos a trabajar con el MemoryRouter */}
                <Route 
                    path="/search"
                    component={SearchScreen} //El componente con el que vamos a trabajar va a ser con el SearchScreen
                />        
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        // expect(wrapper.find('.alert-info').exists()).toBe(true);
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });

    //En esta prueba simulamos que estamos en la ruta que se encuentyra en el url de la pagina Secarch y asegurarnos que la caja de texto tenga ese valor
    test('debe de mostrar a batman y el input con el valor del queryString ', () => {
       
        const wrapper= mount(

            //Como vamos a trabajar con rutas utilizamos el MemoryRouter
            <MemoryRouter initialEntries={['/search?q=batman']}> 
                {/*Trabajamos con rutas porque cuando escribimos algo en el input y le damos enter trabajamos con el url y estar leyendolo, por esa razón vamos a trabajar con el MemoryRouter */}
                <Route 
                    path="/search"
                    component={SearchScreen} //El componente con el que vamos a trabajar va a ser con el SearchScreen
                />        
            </MemoryRouter>
        );

        //Esperamos que el valor que se encuentra en la propiedad value del input contenga batman
        expect( wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar un error si no se muestra el hero', () => {
       
        const wrapper= mount(

            //Como vamos a trabajar con rutas utilizamos el MemoryRouter
            <MemoryRouter initialEntries={['/search?q=batmanfgdfgdgrefd']}> 
                {/*Trabajamos con rutas porque cuando escribimos algo en el input y le damos enter trabajamos con el url y estar leyendolo, por esa razón vamos a trabajar con el MemoryRouter */}
                <Route 
                    path="/search"
                    component={SearchScreen} //El componente con el que vamos a trabajar va a ser con el SearchScreen
                />        
            </MemoryRouter>
        );

        //Esperamos que si existe la clase .alet-danger entonces es true, de lo contrario si no se muestra dicha clase entonces es false
        // expect( wrapper.find('.alert-danger').exists()).toBe(true);
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`There is no a hero with batmanfgdfgdgrefd`);
        expect(wrapper).toMatchSnapshot();
    });

    //Esta prueba es dificil de hacer
    test('debe de llamar el push del history ', () => {
        //****** Lo que hace el formulario es cuando presionamos "Enter" entonces manda la informacion de la caja de texto al url y el url recarga mi componente *********

        // 1.Simulamos el history
        const historyMock = {
            push: jest.fn()
        };

        // 2.Montamos nuestro componente
        const wrapper= mount(

            // 3.Renderizamos el componente
            <MemoryRouter initialEntries={['/search?q=batmanfgdfgdgrefd']}> 
                <Route 
                    path="/search"
                    // 4. Necesitamos enviar el history por lo que hacemos lo siguiente en el component
                    component={ () => <SearchScreen history={historyMock} /> } //Recordemos que el callback () => recibe los props por defecto no es necesario agregarlos y si los quisieramos agregar lo ponemos de la siguiente forma (props)=> solo que no es necesario ya que los recibe por defecto, por lo tanto no es necesario poner los props, recordemos que en este caso la prop que estamos recibiendo del componente SearchScreen es history 
                />       
            </MemoryRouter>
        );

        // ******* Simulamos acciones en el formulario ********

        //5. Simulamos el cambio en la caja de texto
        //Enviamos como primer argumento el evento(change) y como segundo argumento el target
        wrapper.find('input').simulate('change',{
            target:{
                name: 'searchText', //El name de la caja de texto es searchText porque así funciona nuestro customHoon, podemos verlo en el componente SearchScreen
                value:'batman' //Cuando se realiza el push history.push(`?q=${searchText}`); envia este value del input al url. Para entenderlo mejor en este momento ?q=${searchText} es el mismo valor que tenemos en la caja de texto(value)
            }
        });
        
        //6. Simulamos en Onsubmit. Podemos utilizar el prop o el simulate cualquiera de los dos
        //Enviamos como primer argumento el evento onSubmit y recibimos y ejecutamos la función preventDefault
        wrapper.find('form').prop('onSubmit') ( { preventDefault(){} }); //Definimos la función colocando {} al final del preventDefault


        //7. Esperariamos que el historyMock.push haya sido llamado con ?q=${searchText}, cambiamos esto ultimo por el value del input ya que es lo mismo;
        expect( historyMock.push).toHaveBeenCalledWith(`?q=batman`);
    })
    
    
    
})
