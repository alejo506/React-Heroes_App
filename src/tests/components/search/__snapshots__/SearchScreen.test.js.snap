// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar a batman y el input con el valor del queryString  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-10"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        className="d-flex"
        onSubmit={[Function]}
      >
        <input
          aria-label="Search"
          autoComplete="off"
          className="form-control me-2"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="search"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
  </div>
  <h4
    className="mt-4"
  >
     Results 
  </h4>
  <hr />
  <div
    className="row row-cols-auto row-cols-lg-3 g-auto g-lg-3 row-cols-xl-4 g-xl-auto row-cols-xs-2 g-xs-2 animate__animated animate__fadeIn"
  >
    <a
      href="/hero/dc-batman"
      onClick={[Function]}
    >
       
      <div
        className="card my-card "
      >
        <img
          alt="Batman"
          className="card-img"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <h3>
              DC Comics
            </h3>
            <p>
              Primera aparici√≥n: 
              <br />
              Detective Comics #27
            </p>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrar un error si no se muestra el hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-10"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        className="d-flex"
        onSubmit={[Function]}
      >
        <input
          aria-label="Search"
          autoComplete="off"
          className="form-control me-2"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="search"
          value="batmanfgdfgdgrefd"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
  </div>
  <h4
    className="mt-4"
  >
     Results 
  </h4>
  <hr />
  <div
    className="text-center alert alert-danger"
  >
    There is no a hero with 
    <strong>
      batmanfgdfgdgrefd
    </strong>
  </div>
  <div
    className="row row-cols-auto row-cols-lg-3 g-auto g-lg-3 row-cols-xl-4 g-xl-auto row-cols-xs-2 g-xs-2 animate__animated animate__fadeIn"
  />
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-10"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        className="d-flex"
        onSubmit={[Function]}
      >
        <input
          aria-label="Search"
          autoComplete="off"
          className="form-control me-2"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="search"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
  </div>
  <h4
    className="mt-4"
  >
     Results 
  </h4>
  <hr />
  <div
    className="text-center alert alert-info"
  >
    Search a hero
  </div>
  <div
    className="row row-cols-auto row-cols-lg-3 g-auto g-lg-3 row-cols-xl-4 g-xl-auto row-cols-xs-2 g-xs-2 animate__animated animate__fadeIn"
  />
</div>
`;
